src/base-subscription.ts

```diff
@@ -0,0 +1,253 @@
+                    if(err instanceof ErrorResponse && err.statusCode != 204){
+                        this.onError(err);
+                    }
+                    // Because we abort()ed, we will get no more calls to our onreadystatechange handler,
+                    // and so we will not call the event handler again.
+                    // Finish with options.onError instead of the options.onEnd.
+                    
+                } else {
+                    // We consumed some response text, and all's fine. We expect more text.
+                }
+            }
+        }
+        
+        private onDone(): void {
+            if (this.xhr.status === 200) {
+                if (this.state === BaseSubscriptionState.OPENING) {
+                    this.state = BaseSubscriptionState.OPEN;
+                    this.onOpen(responseToHeadersObject(this.xhr.getAllResponseHeaders()));
+                }
+                this.assertStateIsIn( BaseSubscriptionState.OPEN, BaseSubscriptionState.ENDING );
+                let err = this.onChunk();
+                if (err) {
+                    this.state = BaseSubscriptionState.ENDED;
+                    if ( (err as any).statusCode === 204 ) { //TODO: That cast is horrific
+                        this.onEnd();
+                    }
+                    else {
+                        this.onError(err);
+                    }
+                } else if (this.state <= BaseSubscriptionState.ENDING) {
+                    this.onError(new Error("HTTP response ended without receiving EOS message"));
+                } else {
+                    // Stream ended normally.
+                    this.onEnd();
+                }
+            }
+            
+            else {
+                this.assertStateIsIn(BaseSubscriptionState.OPENING, BaseSubscriptionState.OPEN, BaseSubscriptionState.ENDED);
+                
+                if (this.state === BaseSubscriptionState.ENDED) {
+                    // We aborted the request deliberately, and called onError/onEnd elsewhere.
+                    return;
+                }
+                //Something terrible has happened. Most likely a network error. XHR is useless at that point.
+                else if(this.xhr.status === 0) {
+                    this.onError(new NetworkError("Connection lost."));
+                    
+                } else {
+                    this.onError(ErrorResponse.fromXHR(this.xhr));
+                }
+            }
+        }
+        
+        private lastNewlineIndex: number = 0;
+        
+        private onChunk(): Error {
+            this.assertStateIsIn(BaseSubscriptionState.OPEN);
+            let response = this.xhr.responseText;
+            let newlineIndex = response.lastIndexOf("\n");
+            if (newlineIndex > this.lastNewlineIndex) {
+                
+                let rawEvents = response.slice(this.lastNewlineIndex, newlineIndex).split("\n");
```

Comment:
Could you be dropping partial messages here?

Not sure if I'm reading it correctly but it looks like if the chunk had 2.5 messages worth of data in it then you'd just split on the last `"\n"` character and drop the last 0.5 of a message.

Reply:
Added a story to the trello board.